{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **EXAM 2:** JAVIER ARRANZ HERRERO\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "REMEMBER TO PUT THIS CODE IN YOUR TERMINAL\n",
    "\n",
    "* docker start course-mysql\n",
    "* docker stop course-mysql"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Connected: root@mysql'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext sql\n",
    "#%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#First, we need to import all the libraries that we'll need during the exam\n",
    "\n",
    "import re\n",
    "import pymysql.cursors"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ***Problem 1***: Controls\n",
    "\n",
    "Write a Python script that proves that the lines of data in Germplasm.tsv, and LocusGene are in the same sequence, based on the AGI Locus Code (ATxGxxxxxx). (hint: This will help you decide how to load the data into the database)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "For this exercise, we'll need to read the file germplasm and locusgene, and perform a MatchObj to see if both files have the same AGI Locus Code (ATxGxxxxxx). If this is True, this would be our Primary Key to connect both lists.\n",
    "\n",
    "\n",
    "For reading the file, I'll open them first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "germplasmfile = open(\"Germplasm.tsv\", \"r\")\n",
    "locusgenefile = open(\"LocusGene.tsv\", \"r\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once it's opened, I will create a for loop to read everyline (starting from the beggining) looking for the AGI Locus Code (Using MatchObj). But first, we need to tell python that we want to strat from the beggining every time we run the code: #germplasmfile.seek(0)\n",
    "\n",
    "If it finds any AGI Locus in a line, I convert that AGI into a variable called matchObj, and I append the new value of the list to a list that I have created before (**germ = list**).\n",
    "\n",
    "I created a value called ok, which i set it to FALSE.\n",
    "Using this value, I create a new for loop inside the first one and now i look for the same AGI Locus inside the second file, reading everyline.\n",
    "\n",
    "This will compare each line of the first list wht all lines of the second list. With this, we make sure that it will find it even if it is not in the same order.\n",
    "\n",
    "Finally, if the AGI locus found in the first part of the loop is equal to the AGI locus foun in the second part, it changes ok to TRUE and print (MATCH), adding this locus to a different list for locus (**locus = list**).\n",
    "\n",
    "Only in the case it prints MATCH, we can ensure that we have added both locus to both list\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "no match\n",
      "AT1G01040\t MATCH --> AGI added to the lists\n",
      "AT1G01060\t MATCH --> AGI added to the lists\n",
      "AT1G01140\t MATCH --> AGI added to the lists\n",
      "AT1G01220\t MATCH --> AGI added to the lists\n",
      "AT2G03720\t MATCH --> AGI added to the lists\n",
      "AT2G03800\t MATCH --> AGI added to the lists\n",
      "AT2G04240\t MATCH --> AGI added to the lists\n",
      "AT2G05210\t MATCH --> AGI added to the lists\n",
      "AT3G02130\t MATCH --> AGI added to the lists\n",
      "AT3G02140\t MATCH --> AGI added to the lists\n",
      "AT3G02230\t MATCH --> AGI added to the lists\n",
      "AT3G02260\t MATCH --> AGI added to the lists\n",
      "AT3G02310\t MATCH --> AGI added to the lists\n",
      "AT3G02680\t MATCH --> AGI added to the lists\n",
      "AT3G02850\t MATCH --> AGI added to the lists\n",
      "AT3G02870\t MATCH --> AGI added to the lists\n",
      "AT3G03260\t MATCH --> AGI added to the lists\n",
      "AT4G14790\t MATCH --> AGI added to the lists\n",
      "AT4G15210\t MATCH --> AGI added to the lists\n",
      "AT4G15560\t MATCH --> AGI added to the lists\n",
      "AT4G15570\t MATCH --> AGI added to the lists\n",
      "AT4G15802\t MATCH --> AGI added to the lists\n",
      "AT4G15880\t MATCH --> AGI added to the lists\n",
      "AT4G16420\t MATCH --> AGI added to the lists\n",
      "AT4G16480\t MATCH --> AGI added to the lists\n",
      "AT5G10480\t MATCH --> AGI added to the lists\n",
      "AT5G10510\t MATCH --> AGI added to the lists\n",
      "AT5G11110\t MATCH --> AGI added to the lists\n",
      "AT5G11260\t MATCH --> AGI added to the lists\n",
      "AT5G11510\t MATCH --> AGI added to the lists\n",
      "AT5G12200\t MATCH --> AGI added to the lists\n",
      "AT5G13290\t MATCH --> AGI added to the lists\n"
     ]
    }
   ],
   "source": [
    "germplasmfile.seek(0)         #We start from 0 befor the loop\n",
    "\n",
    "germ = []\n",
    "locus = []\n",
    "ok = False\n",
    "\n",
    "for line in germplasmfile.readlines():\n",
    "        \n",
    "    #print(\"the current record is\", line)\n",
    "    matchObj = re.search( r'AT\\dG\\d+\\s', line)\n",
    "    if matchObj:\n",
    "        AGIgerm = matchObj.group()\n",
    "        #print(AGIgerm)\n",
    "        germ.append(AGIgerm)\n",
    "        #print ('AGIgerm added to the list')\n",
    "        locusgenefile.seek(0)         #We start from 0 befor the loop\n",
    "        ok = False\n",
    "        for line2 in locusgenefile.readlines():\n",
    "            matchObj = re.search( r'AT\\dG\\d+\\s', line2)\n",
    "            if matchObj and ok == False:\n",
    "                AGIlocus = matchObj.group()\n",
    "                #print(AGIlocus)\n",
    "                if str(AGIgerm) == str(AGIlocus):\n",
    "                    locus.append(AGIlocus)\n",
    "                    ok = True\n",
    "                    print (AGIlocus, 'MATCH --> AGI added to the lists')\n",
    "                    #print(ok)\n",
    "    else:\n",
    "        print(\"no match\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "NOW we can print both list to see check if they added every value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AT1G01040\\t', 'AT1G01060\\t', 'AT1G01140\\t', 'AT1G01220\\t', 'AT2G03720\\t', 'AT2G03800\\t', 'AT2G04240\\t', 'AT2G05210\\t', 'AT3G02130\\t', 'AT3G02140\\t', 'AT3G02230\\t', 'AT3G02260\\t', 'AT3G02310\\t', 'AT3G02680\\t', 'AT3G02850\\t', 'AT3G02870\\t', 'AT3G03260\\t', 'AT4G14790\\t', 'AT4G15210\\t', 'AT4G15560\\t', 'AT4G15570\\t', 'AT4G15802\\t', 'AT4G15880\\t', 'AT4G16420\\t', 'AT4G16480\\t', 'AT5G10480\\t', 'AT5G10510\\t', 'AT5G11110\\t', 'AT5G11260\\t', 'AT5G11510\\t', 'AT5G12200\\t', 'AT5G13290\\t']\n"
     ]
    }
   ],
   "source": [
    "print(germ)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AT1G01040\\t', 'AT1G01060\\t', 'AT1G01140\\t', 'AT1G01220\\t', 'AT2G03720\\t', 'AT2G03800\\t', 'AT2G04240\\t', 'AT2G05210\\t', 'AT3G02130\\t', 'AT3G02140\\t', 'AT3G02230\\t', 'AT3G02260\\t', 'AT3G02310\\t', 'AT3G02680\\t', 'AT3G02850\\t', 'AT3G02870\\t', 'AT3G03260\\t', 'AT4G14790\\t', 'AT4G15210\\t', 'AT4G15560\\t', 'AT4G15570\\t', 'AT4G15802\\t', 'AT4G15880\\t', 'AT4G16420\\t', 'AT4G16480\\t', 'AT5G10480\\t', 'AT5G10510\\t', 'AT5G11110\\t', 'AT5G11260\\t', 'AT5G11510\\t', 'AT5G12200\\t', 'AT5G13290\\t']\n"
     ]
    }
   ],
   "source": [
    "print(locus)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we close both files and we compare the lists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "germplasmfile.close()\n",
    "locusgenefile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SAME AGI Locus\n"
     ]
    }
   ],
   "source": [
    "if germ == locus:\n",
    "    print('SAME AGI Locus')\n",
    "else:\n",
    "    print('DIFFERENT AGI Locus')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ***Problem 2***: Design and create the database.\n",
    "\n",
    "* It should have two tables - one for each of the two data files.\n",
    "* The two tables should be linked in a 1:1 relationship\n",
    "* you may use either sqlMagic or pymysql to build the database"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now I'm going to create a database with 2 different tables using \"create database 'name'\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql drop database Examdatabase;    #To start from the beginning, drop the database\n",
    "%sql create database Examdatabase;\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "6 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>information_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Examdatabase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>mysql</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>performance_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>sys</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('information_schema',),\n",
       " ('Examdatabase',),\n",
       " ('germplasm',),\n",
       " ('mysql',),\n",
       " ('performance_schema',),\n",
       " ('sys',)]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql show databases #To show the current databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql use Examdatabase; "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the first table of the database, our primary key will be the AGI locus, the rest of columns are text or numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "4 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Field</th>\n",
       "        <th>Type</th>\n",
       "        <th>Null</th>\n",
       "        <th>Key</th>\n",
       "        <th>Default</th>\n",
       "        <th>Extra</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>locus</td>\n",
       "        <td>varchar(15)</td>\n",
       "        <td>NO</td>\n",
       "        <td>PRI</td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm</td>\n",
       "        <td>varchar(40)</td>\n",
       "        <td>YES</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>phenotype</td>\n",
       "        <td>varchar(1000)</td>\n",
       "        <td>YES</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>pubmed</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>YES</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('locus', 'varchar(15)', 'NO', 'PRI', None, ''),\n",
       " ('germplasm', 'varchar(40)', 'YES', '', None, ''),\n",
       " ('phenotype', 'varchar(1000)', 'YES', '', None, ''),\n",
       " ('pubmed', 'int(11)', 'YES', '', None, '')]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#%sql DROP TABLE germplasmtable      #Drop this table to start from 0\n",
    "%sql CREATE TABLE germplasmtable(locus VARCHAR(15) NOT NULL PRIMARY KEY, germplasm VARCHAR(40), phenotype VARCHAR(1000), pubmed int);\n",
    "%sql DESCRIBE germplasmtable\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the other table we use the same Primary Key, this will be our connection. In my case I decided to change the name of the primary to differenciate them, here it will be locusid, containing the AGI Locus."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "3 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Field</th>\n",
       "        <th>Type</th>\n",
       "        <th>Null</th>\n",
       "        <th>Key</th>\n",
       "        <th>Default</th>\n",
       "        <th>Extra</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>locusid</td>\n",
       "        <td>varchar(15)</td>\n",
       "        <td>NO</td>\n",
       "        <td>PRI</td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>gene</td>\n",
       "        <td>varchar(15)</td>\n",
       "        <td>YES</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>proteinlength</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>YES</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('locusid', 'varchar(15)', 'NO', 'PRI', None, ''),\n",
       " ('gene', 'varchar(15)', 'YES', '', None, ''),\n",
       " ('proteinlength', 'int(11)', 'YES', '', None, '')]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#%sql DROP TABLE locusgenetable\n",
    "%sql CREATE TABLE locusgenetable(locusid VARCHAR(15) NOT NULL PRIMARY KEY, gene VARCHAR(15), proteinlength int);\n",
    "%sql DESCRIBE locusgenetable"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see the content of the tables using the command SELECT\n",
    "\n",
    "(REMEMBER: we'll see only the last untagged SELECT, change the tag from one to another)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>locus</th>\n",
       "        <th>germplasm</th>\n",
       "        <th>phenotype</th>\n",
       "        <th>pubmed</th>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT * FROM germplasmtable;\n",
    "#%sql SELECT * FROM locusgenetable;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "AT THIS MOMENT, we already have our database with our tables **germpasmtable** and **locusgenetable** created"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ***Problem 3***: Fill the database.\n",
    "\n",
    "Using pymysql, create a Python script that reads the data from these files, and fills the database. There are a variety of strategies to accomplish this. I will give all strategies equal credit - do whichever one you are most confident with.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "PRINTING GERMPLASM TABLE\n",
      "\n",
      "\n",
      "\n",
      "INSERT FAILED\n",
      "written Locus germplasm phenotype pubmed\n",
      "'\n",
      "written AT1G01040 CS3828 Increased abundance of miRNA precursors. 17369351\n",
      "'\n",
      "written AT1G01060 lhy-101 The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant. 16891401\n",
      "'\n",
      "written AT1G01140 SALK_058629 hypersensitive to low potassium media 17486125\n",
      "'\n",
      "written AT1G01220 SALK_012400C fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected. 18199744\n",
      "'\n",
      "written AT2G03720 SALK_042433 Multiple straight hairs 16367956\n",
      "'\n",
      "written AT2G03800 gek1-1 Ethanol hypersensitivity. 15215505\n",
      "'\n",
      "written AT2G04240 xerico Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype. 17933900\n",
      "'\n",
      "written AT2G05210 pot1-1 No visible phenotype. 17627276\n",
      "'\n",
      "written AT3G02130 rpk2-2 The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants. 17419837\n",
      "'\n",
      "written AT3G02140 afp4-1 Decreased germination on high concentrations of glucose and sorbitol. 18484180\n",
      "'\n",
      "written AT3G02230 rgp1-1 rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls. 21478444\n",
      "'\n",
      "written AT3G02260 tir3-1 RGLG1:rglg1 rglg2 The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence. 17586653\n",
      "'\n",
      "written AT3G02310 sep2-1 Non-described subtle phenotype. 10821278\n",
      "'\n",
      "written AT3G02680 atnbs1-1 Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions. 17672843\n",
      "'\n",
      "written AT3G02850 CS3816 The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion. 17568770\n",
      "'\n",
      "written AT3G02870 vtc4-1 ascorbate deficient 16595667\n",
      "'\n",
      "written AT3G03260 hdg8-1 No visible phenotype. 16778018\n",
      "'\n",
      "written AT4G14790 pdd17 Defective pollen development. 19237690\n",
      "'\n",
      "written AT4G15210 bmy1-2 Plants cold-shocked for 6h have an increased starch content compared to wildtype. 16297066\n",
      "'\n",
      "written AT4G15560 cla1-1 Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves. 10982425\n",
      "'\n",
      "written AT4G15570 maa3 Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT. 18772186\n",
      "'\n",
      "written AT4G15802 Athspb-2 Early flowering, reduced fertility, aborted seeds. 20388662\n",
      "'\n",
      "written AT4G15880 esd4-2 Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1. 17513499\n",
      "'\n",
      "written AT4G16420 prz1-1 Altered response to auxin and cytokinin 12747832\n",
      "'\n",
      "written AT4G16480 atint4-2 No visible phenotype. 16603666\n",
      "'\n",
      "written AT5G10480 pas2-3 Segregates 25% embryo lethal. 18799749\n",
      "'\n",
      "written AT5G10510 plt3-1 Short roots and shortened root meristem. 17960244\n",
      "'\n",
      "written AT5G11110 kns2 Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile. 18779216\n",
      "'\n",
      "written AT5G11260 hy5-101 Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller. 16891401\n",
      "'\n",
      "written AT5G11510 myb3r4-1 No visible phenotype. 17287251\n",
      "'\n",
      "written AT5G12200 pyd2-2 The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source. 19413687\n",
      "'\n",
      "written AT5G13290 crn-1 Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3. 12345678\n",
      "'\n",
      "\n",
      "\n",
      "\n",
      "PRINTING LOCUSGENE TABLE\n",
      "\n",
      "\n",
      "\n",
      "INSERT FAILED\n",
      "written Locus Gene ProteinLength\n",
      "'\n",
      "written AT1G01040 DCL1 332\n",
      "'\n",
      "written AT1G01060 LHY 290\n",
      "'\n",
      "written AT1G01140 CIPK9 223\n",
      "'\n",
      "written AT1G01220 FKGP 190\n",
      "'\n",
      "written AT2G03720 MRH6 189\n",
      "'\n",
      "written AT2G03800 GEK1 196\n",
      "'\n",
      "written AT2G04240 XERICO 256\n",
      "'\n",
      "written AT2G05210 POT1A 221\n",
      "'\n",
      "written AT3G02130 RPK2 284\n",
      "'\n",
      "written AT3G02140 TMAC2 300\n",
      "'\n",
      "written AT3G02230 RGP1 301\n",
      "'\n",
      "written AT3G02260 BIG 279\n",
      "'\n",
      "written AT3G02310 SEP2 175\n",
      "'\n",
      "written AT3G02680 NBS1 190\n",
      "'\n",
      "written AT3G02850 SKOR 234\n",
      "'\n",
      "written AT3G02870 VTC4 311\n",
      "'\n",
      "written AT3G03260 HDG8 194\n",
      "'\n",
      "written AT4G14790 SUV3 312\n",
      "'\n",
      "written AT4G15210 BAM5 313\n",
      "'\n",
      "written AT4G15560 DXS 219\n",
      "'\n",
      "written AT4G15570 MAA3 294\n",
      "'\n",
      "written AT4G15802 HSBP 254\n",
      "'\n",
      "written AT4G15880 ESD4 265\n",
      "'\n",
      "written AT4G16420 ADA2B 279\n",
      "'\n",
      "written AT4G16480 INT4 284\n",
      "'\n",
      "written AT5G10480 PAS2 301\n",
      "'\n",
      "written AT5G10510 AIL6 310\n",
      "'\n",
      "written AT5G11110 SPS2 232\n",
      "'\n",
      "written AT5G11260 HY5 221\n",
      "'\n",
      "written AT5G11510 MYB3R-4 336\n",
      "'\n",
      "written AT5G12200 PYD2 310\n",
      "'\n",
      "written AT5G13290 CRN 189\n",
      "'\n"
     ]
    }
   ],
   "source": [
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='Examdatabase',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "#connection.autocommit = False\n",
    "\n",
    "\n",
    "def write_to_database(db, locus, germplasm, phenotype, pubmed):\n",
    "    locus = str(locus)\n",
    "    germplasm = str(germplasm)\n",
    "    phenotype = str(phenotype)\n",
    "    pubmed = str(pubmed)\n",
    "    \n",
    "    try:\n",
    "        with db.cursor() as cursor:\n",
    "            sql = \"insert into germplasmtable (locus, germplasm, phenotype, pubmed) values ('{}','{}', '{}', '{}')\".format(locus, germplasm, phenotype, pubmed)\n",
    "            cursor.execute(sql)\n",
    "            \n",
    "            db.commit()\n",
    "    except Exception as e:\n",
    "        print(\"INSERT FAILED\")\n",
    "        return False\n",
    "    finally:\n",
    "        return True\n",
    "    \n",
    "    \n",
    "\n",
    "def write_to_database2(db, locusid, gene, proteinlength):\n",
    "    locusid = str(locusid)\n",
    "    gene = str(gene)\n",
    "    proteinlength = str(proteinlength)\n",
    "    \n",
    "    try:\n",
    "        with db.cursor() as cursor:\n",
    "            sql = \"insert into locusgenetable (locusid, gene, proteinlength) values ('{}','{}', '{}')\".format(locusid, gene, proteinlength)\n",
    "            cursor.execute(sql)\n",
    "            \n",
    "            db.commit()\n",
    "    except Exception as e:\n",
    "        print(\"INSERT FAILED\")\n",
    "        return False\n",
    "    finally:\n",
    "        return True\n",
    "    \n",
    "        \n",
    "germplasmfile = open(\"Germplasm.tsv\", \"r\")\n",
    "locusgenefile = open(\"LocusGene.tsv\", \"r\")\n",
    "germplasmfile.seek(0)         #We start from 0 befor the loop\n",
    "locusgenefile.seek(0)\n",
    "\n",
    "print(\"\")\n",
    "print(\"\")\n",
    "print(\"\")\n",
    "print(\"PRINTING GERMPLASM TABLE\")\n",
    "print(\"\")\n",
    "print(\"\")\n",
    "print(\"\")\n",
    "\n",
    "for germline in germplasmfile.readlines():\n",
    "    locus, germplasm, phenotype, pubmed = germline.split('\\t')\n",
    "    if write_to_database(connection, locus, germplasm, phenotype, pubmed):\n",
    "        print(\"written {} {} {} {}'\".format(locus, germplasm, phenotype, pubmed))\n",
    "    else:\n",
    "        print(\"write failed\")\n",
    "\n",
    "    \n",
    "\n",
    "print(\"\")\n",
    "print(\"\")\n",
    "print(\"\")\n",
    "print(\"PRINTING LOCUSGENE TABLE\")\n",
    "print(\"\")\n",
    "print(\"\")\n",
    "print(\"\")\n",
    "    \n",
    "    \n",
    "    \n",
    "germplasmfile.close()\n",
    "#locusgenefile.close()\n",
    "\n",
    "\n",
    "for locusline in locusgenefile.readlines():\n",
    "    locusid, gene, proteinlength = locusline.split('\\t')\n",
    "    if write_to_database2(connection, locusid, gene, proteinlength):\n",
    "        print(\"written {} {} {}'\".format(locusid, gene, proteinlength))\n",
    "    else:\n",
    "        print(\"write failed\")\n",
    "        \n",
    "        \n",
    "locusgenefile.close()\n",
    "\n",
    "connection.close()\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4: Create reports, written to a file\n",
    "\n",
    "1. Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "\n",
    "2. Create a joined report that only includes the Genes SKOR and MAA3\n",
    "\n",
    "3. Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "4. Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "When creating reports 2 and 3, remember the \"Don't Repeat Yourself\" rule! \n",
    "\n",
    "All reports should be written to **the same file**.  You may name the file anything you wish."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lets see how we inserted the data in the table using SELECT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%sql SELECT * FROM germplasmtable;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%sql SELECT * FROM locusgenetable;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ONCE WE CHECK out data is correct, we can continue with the problem 4.\n",
    "\n",
    "(**NOTE:** I have tagged the %sql SELECT to reduce space in the exam, delete \"#\" to see the tables)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the NEXT problem, I selected only de locusgenetable primary key (locusgenetable.locusid), and all the columns that I want in my report. Then, I use a LEFT JOIN to include in the table the columns that we need, instead of creating a table that contains the AGI locus 2 times.\n",
    "\n",
    "FInally I said that the primary key in both is AGI LOCUS (locusgenetable.locusid = germplasmtable.locus)\n",
    "\n",
    "(**IMPORTANT NOTE: I WILL CREATE THE WHOLE REPORT AT THE END, ONCE I HAVE THE CODE FOR EACH PART.**)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "32 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>locusid</th>\n",
       "        <th>gene</th>\n",
       "        <th>proteinlength</th>\n",
       "        <th>germplasm</th>\n",
       "        <th>phenotype</th>\n",
       "        <th>pubmed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01040</td>\n",
       "        <td>DCL1</td>\n",
       "        <td>332</td>\n",
       "        <td>CS3828</td>\n",
       "        <td>Increased abundance of miRNA precursors.</td>\n",
       "        <td>17369351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01060</td>\n",
       "        <td>LHY</td>\n",
       "        <td>290</td>\n",
       "        <td>lhy-101</td>\n",
       "        <td>The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.</td>\n",
       "        <td>16891401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01140</td>\n",
       "        <td>CIPK9</td>\n",
       "        <td>223</td>\n",
       "        <td>SALK_058629</td>\n",
       "        <td>hypersensitive to low potassium media</td>\n",
       "        <td>17486125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01220</td>\n",
       "        <td>FKGP</td>\n",
       "        <td>190</td>\n",
       "        <td>SALK_012400C</td>\n",
       "        <td>fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.</td>\n",
       "        <td>18199744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03720</td>\n",
       "        <td>MRH6</td>\n",
       "        <td>189</td>\n",
       "        <td>SALK_042433</td>\n",
       "        <td>Multiple straight hairs</td>\n",
       "        <td>16367956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03800</td>\n",
       "        <td>GEK1</td>\n",
       "        <td>196</td>\n",
       "        <td>gek1-1</td>\n",
       "        <td>Ethanol hypersensitivity.</td>\n",
       "        <td>15215505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G04240</td>\n",
       "        <td>XERICO</td>\n",
       "        <td>256</td>\n",
       "        <td>xerico</td>\n",
       "        <td>Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.</td>\n",
       "        <td>17933900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G05210</td>\n",
       "        <td>POT1A</td>\n",
       "        <td>221</td>\n",
       "        <td>pot1-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>17627276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02130</td>\n",
       "        <td>RPK2</td>\n",
       "        <td>284</td>\n",
       "        <td>rpk2-2</td>\n",
       "        <td>The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.</td>\n",
       "        <td>17419837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02140</td>\n",
       "        <td>TMAC2</td>\n",
       "        <td>300</td>\n",
       "        <td>afp4-1</td>\n",
       "        <td>Decreased germination on high concentrations of glucose and sorbitol.</td>\n",
       "        <td>18484180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02230</td>\n",
       "        <td>RGP1</td>\n",
       "        <td>301</td>\n",
       "        <td>rgp1-1</td>\n",
       "        <td>rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.</td>\n",
       "        <td>21478444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02260</td>\n",
       "        <td>BIG</td>\n",
       "        <td>279</td>\n",
       "        <td>tir3-1 RGLG1:rglg1 rglg2</td>\n",
       "        <td>The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.</td>\n",
       "        <td>17586653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02310</td>\n",
       "        <td>SEP2</td>\n",
       "        <td>175</td>\n",
       "        <td>sep2-1</td>\n",
       "        <td>Non-described subtle phenotype.</td>\n",
       "        <td>10821278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02680</td>\n",
       "        <td>NBS1</td>\n",
       "        <td>190</td>\n",
       "        <td>atnbs1-1</td>\n",
       "        <td>Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.</td>\n",
       "        <td>17672843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>SKOR</td>\n",
       "        <td>234</td>\n",
       "        <td>CS3816</td>\n",
       "        <td>The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.</td>\n",
       "        <td>17568770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02870</td>\n",
       "        <td>VTC4</td>\n",
       "        <td>311</td>\n",
       "        <td>vtc4-1</td>\n",
       "        <td>ascorbate deficient</td>\n",
       "        <td>16595667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G03260</td>\n",
       "        <td>HDG8</td>\n",
       "        <td>194</td>\n",
       "        <td>hdg8-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>16778018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G14790</td>\n",
       "        <td>SUV3</td>\n",
       "        <td>312</td>\n",
       "        <td>pdd17</td>\n",
       "        <td>Defective pollen development.</td>\n",
       "        <td>19237690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15210</td>\n",
       "        <td>BAM5</td>\n",
       "        <td>313</td>\n",
       "        <td>bmy1-2</td>\n",
       "        <td>Plants cold-shocked for 6h have an increased starch content compared to wildtype.</td>\n",
       "        <td>16297066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15560</td>\n",
       "        <td>DXS</td>\n",
       "        <td>219</td>\n",
       "        <td>cla1-1</td>\n",
       "        <td>Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.</td>\n",
       "        <td>10982425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>MAA3</td>\n",
       "        <td>294</td>\n",
       "        <td>maa3</td>\n",
       "        <td>Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.</td>\n",
       "        <td>18772186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15802</td>\n",
       "        <td>HSBP</td>\n",
       "        <td>254</td>\n",
       "        <td>Athspb-2</td>\n",
       "        <td>Early flowering, reduced fertility, aborted seeds.</td>\n",
       "        <td>20388662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15880</td>\n",
       "        <td>ESD4</td>\n",
       "        <td>265</td>\n",
       "        <td>esd4-2</td>\n",
       "        <td>Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.</td>\n",
       "        <td>17513499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16420</td>\n",
       "        <td>ADA2B</td>\n",
       "        <td>279</td>\n",
       "        <td>prz1-1</td>\n",
       "        <td>Altered response to auxin and cytokinin</td>\n",
       "        <td>12747832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16480</td>\n",
       "        <td>INT4</td>\n",
       "        <td>284</td>\n",
       "        <td>atint4-2</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>16603666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10480</td>\n",
       "        <td>PAS2</td>\n",
       "        <td>301</td>\n",
       "        <td>pas2-3</td>\n",
       "        <td>Segregates 25% embryo lethal.</td>\n",
       "        <td>18799749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10510</td>\n",
       "        <td>AIL6</td>\n",
       "        <td>310</td>\n",
       "        <td>plt3-1</td>\n",
       "        <td>Short roots and shortened root meristem.</td>\n",
       "        <td>17960244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11110</td>\n",
       "        <td>SPS2</td>\n",
       "        <td>232</td>\n",
       "        <td>kns2</td>\n",
       "        <td>Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.</td>\n",
       "        <td>18779216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11260</td>\n",
       "        <td>HY5</td>\n",
       "        <td>221</td>\n",
       "        <td>hy5-101</td>\n",
       "        <td>Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.</td>\n",
       "        <td>16891401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11510</td>\n",
       "        <td>MYB3R-4</td>\n",
       "        <td>336</td>\n",
       "        <td>myb3r4-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>17287251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G12200</td>\n",
       "        <td>PYD2</td>\n",
       "        <td>310</td>\n",
       "        <td>pyd2-2</td>\n",
       "        <td>The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.</td>\n",
       "        <td>19413687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G13290</td>\n",
       "        <td>CRN</td>\n",
       "        <td>189</td>\n",
       "        <td>crn-1</td>\n",
       "        <td>Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.</td>\n",
       "        <td>12345678</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AT1G01040', 'DCL1', 332, 'CS3828', 'Increased abundance of miRNA precursors.', 17369351),\n",
       " ('AT1G01060', 'LHY', 290, 'lhy-101', 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', 16891401),\n",
       " ('AT1G01140', 'CIPK9', 223, 'SALK_058629', 'hypersensitive to low potassium media', 17486125),\n",
       " ('AT1G01220', 'FKGP', 190, 'SALK_012400C', 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ s ... (41 characters truncated) ... henotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', 18199744),\n",
       " ('AT2G03720', 'MRH6', 189, 'SALK_042433', 'Multiple straight hairs', 16367956),\n",
       " ('AT2G03800', 'GEK1', 196, 'gek1-1', 'Ethanol hypersensitivity.', 15215505),\n",
       " ('AT2G04240', 'XERICO', 256, 'xerico', 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', 17933900),\n",
       " ('AT2G05210', 'POT1A', 221, 'pot1-1', 'No visible phenotype.', 17627276),\n",
       " ('AT3G02130', 'RPK2', 284, 'rpk2-2', 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', 17419837),\n",
       " ('AT3G02140', 'TMAC2', 300, 'afp4-1', 'Decreased germination on high concentrations of glucose and sorbitol.', 18484180),\n",
       " ('AT3G02230', 'RGP1', 301, 'rgp1-1', 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', 21478444),\n",
       " ('AT3G02260', 'BIG', 279, 'tir3-1 RGLG1:rglg1 rglg2', 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', 17586653),\n",
       " ('AT3G02310', 'SEP2', 175, 'sep2-1', 'Non-described subtle phenotype.', 10821278),\n",
       " ('AT3G02680', 'NBS1', 190, 'atnbs1-1', 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', 17672843),\n",
       " ('AT3G02850', 'SKOR', 234, 'CS3816', 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 17568770),\n",
       " ('AT3G02870', 'VTC4', 311, 'vtc4-1', 'ascorbate deficient', 16595667),\n",
       " ('AT3G03260', 'HDG8', 194, 'hdg8-1', 'No visible phenotype.', 16778018),\n",
       " ('AT4G14790', 'SUV3', 312, 'pdd17', 'Defective pollen development.', 19237690),\n",
       " ('AT4G15210', 'BAM5', 313, 'bmy1-2', 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', 16297066),\n",
       " ('AT4G15560', 'DXS', 219, 'cla1-1', 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', 10982425),\n",
       " ('AT4G15570', 'MAA3', 294, 'maa3', 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 18772186),\n",
       " ('AT4G15802', 'HSBP', 254, 'Athspb-2', 'Early flowering, reduced fertility, aborted seeds.', 20388662),\n",
       " ('AT4G15880', 'ESD4', 265, 'esd4-2', 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', 17513499),\n",
       " ('AT4G16420', 'ADA2B', 279, 'prz1-1', 'Altered response to auxin and cytokinin', 12747832),\n",
       " ('AT4G16480', 'INT4', 284, 'atint4-2', 'No visible phenotype.', 16603666),\n",
       " ('AT5G10480', 'PAS2', 301, 'pas2-3', 'Segregates 25% embryo lethal.', 18799749),\n",
       " ('AT5G10510', 'AIL6', 310, 'plt3-1', 'Short roots and shortened root meristem.', 17960244),\n",
       " ('AT5G11110', 'SPS2', 232, 'kns2', 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', 18779216),\n",
       " ('AT5G11260', 'HY5', 221, 'hy5-101', 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', 16891401),\n",
       " ('AT5G11510', 'MYB3R-4', 336, 'myb3r4-1', 'No visible phenotype.', 17287251),\n",
       " ('AT5G12200', 'PYD2', 310, 'pyd2-2', 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', 19413687),\n",
       " ('AT5G13290', 'CRN', 189, 'crn-1', 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', 12345678)]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT locusgenetable.locusid, gene, proteinlength, germplasm, phenotype, pubmed FROM locusgenetable LEFT JOIN germplasmtable ON locusgenetable.locusid = germplasmtable.locus;\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the second part I do exactly the same (%sql SELECT). The main difference is that I will use the command WHERE to see only the lines that contain the word \"SKOR\" in the column gene (**WHERE gene LIKE 'SKOR'**) and the command OR to print also if it finds **'MAA3'**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "2 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>locusid</th>\n",
       "        <th>gene</th>\n",
       "        <th>proteinlength</th>\n",
       "        <th>germplasm</th>\n",
       "        <th>phenotype</th>\n",
       "        <th>pubmed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>SKOR</td>\n",
       "        <td>234</td>\n",
       "        <td>CS3816</td>\n",
       "        <td>The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.</td>\n",
       "        <td>17568770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>MAA3</td>\n",
       "        <td>294</td>\n",
       "        <td>maa3</td>\n",
       "        <td>Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.</td>\n",
       "        <td>18772186</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AT3G02850', 'SKOR', 234, 'CS3816', 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 17568770),\n",
       " ('AT4G15570', 'MAA3', 294, 'maa3', 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 18772186)]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT locusgenetable.locusid, gene, proteinlength, germplasm, phenotype, pubmed FROM locusgenetable LEFT JOIN germplasmtable ON locusgenetable.locusid = germplasmtable.locus WHERE gene LIKE 'SKOR' OR gene LIKE 'MAA3';"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The thir point says that we want to count all the lines. For this aim we'll use the command SELECT COUNT.\n",
    "As we want to select everythin (all genes from AG1 to AG5) we write a star (*).\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Number Of Matches</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>32</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[(32,)]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT COUNT(*) AS \"Number Of Matches\" FROM locusgenetable;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the Average of the protein length we can use the command SELECT AVG(column) and we select the proteinlength column from the locusgenetable TABLE."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>AVG(proteinlength)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>258.8750</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[(Decimal('258.8750'),)]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT AVG(proteinlength) FROM locusgenetable;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "FINALLY, we want to write a **FULL SINGLE REPORT** with all the reports that we have created before. For this, I will convert all our reports into variables. Those variables we want them to be just text, so we use the STING command.\n",
    "\n",
    "Last, we open a new file that I created in the console with the command **TOUCH EXAMFILEREPORT.csv and TOUCH EXAMFILEREPORT.txt** (the txt file is just to compare the difference). We need to open it to WRITE on it (\"w\").\n",
    "\n",
    "Then we write our variables separated by a couple of new lines (\\n) and a title that I gave them to see which part corresponds to which report.\n",
    "\n",
    "As the file writes and deletes what there as before in the file, we want to use the .wrte command only 1 time (that is why we create the variables). In addition, we use the curve brackets \"{}\" to introduce our variable in the text we're writing in the file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "32 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "2 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      "FULL REPORT +-----------+---------+---------------+--------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+\n",
      "|  locusid  |   gene  | proteinlength |        germplasm         |                                                                                                                                                                      phenotype                                                                                                                                                                      |  pubmed  |\n",
      "+-----------+---------+---------------+--------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+\n",
      "| AT1G01040 |   DCL1  |      332      |          CS3828          |                                                                                                                                                       Increased abundance of miRNA precursors.                                                                                                                                                      | 17369351 |\n",
      "| AT1G01060 |   LHY   |      290      |         lhy-101          |                                                                        The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.                                                                        | 16891401 |\n",
      "| AT1G01140 |  CIPK9  |      223      |       SALK_058629        |                                                                                                                                                        hypersensitive to low potassium media                                                                                                                                                        | 17486125 |\n",
      "| AT1G01220 |   FKGP  |      190      |       SALK_012400C       | fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected. | 18199744 |\n",
      "| AT2G03720 |   MRH6  |      189      |       SALK_042433        |                                                                                                                                                               Multiple straight hairs                                                                                                                                                               | 16367956 |\n",
      "| AT2G03800 |   GEK1  |      196      |          gek1-1          |                                                                                                                                                              Ethanol hypersensitivity.                                                                                                                                                              | 15215505 |\n",
      "| AT2G04240 |  XERICO |      256      |          xerico          |                                                                                                                              Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.                                                                                                                             | 17933900 |\n",
      "| AT2G05210 |  POT1A  |      221      |          pot1-1          |                                                                                                                                                                No visible phenotype.                                                                                                                                                                | 17627276 |\n",
      "| AT3G02130 |   RPK2  |      284      |          rpk2-2          |                             The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.                             | 17419837 |\n",
      "| AT3G02140 |  TMAC2  |      300      |          afp4-1          |                                                                                                                                        Decreased germination on high concentrations of glucose and sorbitol.                                                                                                                                        | 18484180 |\n",
      "| AT3G02230 |   RGP1  |      301      |          rgp1-1          |                                                                                   rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.                                                                                   | 21478444 |\n",
      "| AT3G02260 |   BIG   |      279      | tir3-1 RGLG1:rglg1 rglg2 |                                                                                                                The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.                                                                                                                | 17586653 |\n",
      "| AT3G02310 |   SEP2  |      175      |          sep2-1          |                                                                                                                                                           Non-described subtle phenotype.                                                                                                                                                           | 10821278 |\n",
      "| AT3G02680 |   NBS1  |      190      |         atnbs1-1         |                                                                                                 Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.                                                                                                | 17672843 |\n",
      "| AT3G02850 |   SKOR  |      234      |          CS3816          |                                                                                                                                     The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.                                                                                                                                    | 17568770 |\n",
      "| AT3G02870 |   VTC4  |      311      |          vtc4-1          |                                                                                                                                                                 ascorbate deficient                                                                                                                                                                 | 16595667 |\n",
      "| AT3G03260 |   HDG8  |      194      |          hdg8-1          |                                                                                                                                                                No visible phenotype.                                                                                                                                                                | 16778018 |\n",
      "| AT4G14790 |   SUV3  |      312      |          pdd17           |                                                                                                                                                            Defective pollen development.                                                                                                                                                            | 19237690 |\n",
      "| AT4G15210 |   BAM5  |      313      |          bmy1-2          |                                                                                                                                  Plants cold-shocked for 6h have an increased starch content compared to wildtype.                                                                                                                                  | 16297066 |\n",
      "| AT4G15560 |   DXS   |      219      |          cla1-1          |                                                                                                              Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.                                                                                                              | 10982425 |\n",
      "| AT4G15570 |   MAA3  |      294      |           maa3           |                                                                          Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.                                                                         | 18772186 |\n",
      "| AT4G15802 |   HSBP  |      254      |         Athspb-2         |                                                                                                                                                  Early flowering, reduced fertility, aborted seeds.                                                                                                                                                 | 20388662 |\n",
      "| AT4G15880 |   ESD4  |      265      |          esd4-2          |                                                                                                             Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.                                                                                                             | 17513499 |\n",
      "| AT4G16420 |  ADA2B  |      279      |          prz1-1          |                                                                                                                                                       Altered response to auxin and cytokinin                                                                                                                                                       | 12747832 |\n",
      "| AT4G16480 |   INT4  |      284      |         atint4-2         |                                                                                                                                                                No visible phenotype.                                                                                                                                                                | 16603666 |\n",
      "| AT5G10480 |   PAS2  |      301      |          pas2-3          |                                                                                                                                                            Segregates 25% embryo lethal.                                                                                                                                                            | 18799749 |\n",
      "| AT5G10510 |   AIL6  |      310      |          plt3-1          |                                                                                                                                                       Short roots and shortened root meristem.                                                                                                                                                      | 17960244 |\n",
      "| AT5G11110 |   SPS2  |      232      |           kns2           |                                                                                                        Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.                                                                                                        | 18779216 |\n",
      "| AT5G11260 |   HY5   |      221      |         hy5-101          |                                                                        Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.                                                                        | 16891401 |\n",
      "| AT5G11510 | MYB3R-4 |      336      |         myb3r4-1         |                                                                                                                                                                No visible phenotype.                                                                                                                                                                | 17287251 |\n",
      "| AT5G12200 |   PYD2  |      310      |          pyd2-2          |                                     The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.                                     | 19413687 |\n",
      "| AT5G13290 |   CRN   |      189      |          crn-1           |                                                              Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.                                                              | 12345678 |\n",
      "+-----------+---------+---------------+--------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+\n",
      "\n",
      "\n",
      "\n",
      "SKOR AND MAA3+-----------+------+---------------+-----------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+\n",
      "|  locusid  | gene | proteinlength | germplasm |                                                                                             phenotype                                                                                              |  pubmed  |\n",
      "+-----------+------+---------------+-----------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+\n",
      "| AT3G02850 | SKOR |      234      |   CS3816  |                                                            The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.                                                            | 17568770 |\n",
      "| AT4G15570 | MAA3 |      294      |    maa3   | Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT. | 18772186 |\n",
      "+-----------+------+---------------+-----------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+\n",
      "\n",
      "\n",
      "\n",
      "Number Of Entries+-------------------+\n",
      "| Number Of Matches |\n",
      "+-------------------+\n",
      "|         32        |\n",
      "+-------------------+\n",
      "\n",
      "\n",
      "\n",
      "Average Protein Length+--------------------+\n",
      "| AVG(proteinlength) |\n",
      "+--------------------+\n",
      "|      258.8750      |\n",
      "+--------------------+\n"
     ]
    }
   ],
   "source": [
    "report1 = %sql SELECT locusgenetable.locusid, gene, proteinlength, germplasm, phenotype, pubmed FROM locusgenetable LEFT JOIN germplasmtable ON locusgenetable.locusid = germplasmtable.locus;\n",
    "#print(report1)\n",
    "report2 = %sql SELECT locusgenetable.locusid, gene, proteinlength, germplasm, phenotype, pubmed FROM locusgenetable LEFT JOIN germplasmtable ON locusgenetable.locusid = germplasmtable.locus WHERE gene LIKE 'SKOR' OR gene LIKE 'MAA3';\n",
    "#print(report2)\n",
    "report3 = %sql SELECT COUNT(*) AS \"Number Of Matches\" FROM locusgenetable;\n",
    "#print(report3)\n",
    "report4 = %sql SELECT AVG(proteinlength) FROM locusgenetable;\n",
    "#print(report4)\n",
    "\n",
    "\n",
    "report1= str(report1)\n",
    "report2= str(report2)\n",
    "report3= str(report3)\n",
    "report4= str(report4)\n",
    "\n",
    "FullReport = open(\"REPORT_Exam2.csv\", \"w\")\n",
    "\n",
    "FullReport.write(\"FULL REPORT {}\\n\\n\\n\\nSKOR AND MAA3{}\\n\\n\\n\\nNumber Of Entries{}\\n\\n\\n\\nAverage Protein Length{}\".format(report1, report2, report3, report4)) \n",
    "\n",
    "FullReport.close()\n",
    "\n",
    "\n",
    "checkcontent = open(\"REPORT_Exam2.csv\", \"r\")\n",
    "print(checkcontent.read())  # print the content of the file\n",
    "#print('written')\n",
    "checkcontent.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
